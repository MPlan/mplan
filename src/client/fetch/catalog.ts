import { fetchWithAuth } from './fetch-with-auth';
import * as Model from 'models';

export async function fetchCatalog() {
  const response = await fetchWithAuth('/api/catalog');
  const catalog = (await response.json()) as Model.Catalog.Model;
  return catalog;
}
