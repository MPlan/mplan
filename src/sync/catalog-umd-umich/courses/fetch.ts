import axios from 'axios';

export async function fetchCourseList(level: 'undergraduate' | 'graduate', subjectCode: string) {
  const response = await axios({
    url: `http://catalog.umd.umich.edu/${level}/coursesaz/${subjectCode.toLowerCase().trim()}`,
    headers: {
      'X-What-Is-This-Request': 'https://github.com/MPlan/mplan',
    },
  });
  const html = response.data as string | undefined;
  if (!html) throw new Error(`could not get course list for "${level} ${subjectCode}"`);
  return html;
}
