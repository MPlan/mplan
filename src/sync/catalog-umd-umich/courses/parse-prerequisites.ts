import { oneLine } from 'common-tags';
import { formDecode } from 'utilities/form-encode-decode';
import { Prerequisite } from '../models';

/** A simple interface that extends an array */
export interface ParseTree extends Array<string | ParseTree> {
  [key: number]: string | ParseTree;
}

/**
 * Given the HTML containing the prerequisites, this function will find all the anchors denoting
 * course in the catalog and replace them with a `__SUBJECT-CODE|COURSE-NUMBER__` directive. This
 * directive will be parsed out later.
 */
export function replacePrerequisiteAnchors(courseBlockExtra: Element) {
  const document = courseBlockExtra.ownerDocument;
  const anchors = Array.from(courseBlockExtra.querySelectorAll('a'))
    .filter(link => {
      const query = link.href.split('?')[1];
      if (!query) return false;
      const decoded = formDecode(query);
      if (!decoded.P) return false;
      const p = decoded.P;
      if (!/(\w*)\s*(\w*)/.test(p)) return false;
      return true;
    })
    .map(anchor => {
      const query = anchor.href.split('?')[1];
      const decoded = formDecode(query);
      const p = decoded.P;
      const subjectCodeCourseNumberMatch = /(\w*)\s*(\w*)/i.exec(p);
      if (!subjectCodeCourseNumberMatch) {
        throw new Error(
          'could get not course and subject code from anchor when parsing prerequisites',
        );
      }
      const subjectCode = subjectCodeCourseNumberMatch[1];
      const courseNumber = subjectCodeCourseNumberMatch[2];
      if (!subjectCode) throw new Error('subject code was falsy when parsing prerequisites');
      if (!courseNumber) throw new Error('course number was falsy when parsing prerequisites');

      return {
        anchor,
        subjectCode,
        courseNumber,
      };
    });

  for (const { anchor, subjectCode, courseNumber } of anchors) {
    const anchorParent = anchor.parentElement;
    if (!anchorParent) {
      throw new Error(`Anchor parent was null`); // should never happen
    }
    // replaces the anchor with an easily parsed string directive
    anchorParent.replaceChild(
      document.createTextNode(`__${subjectCode}|${courseNumber}__`),
      anchor,
    );
  }

  const arr = [courseBlockExtra.textContent || ''];
  const oneLined = oneLine(Object.assign(arr, { raw: arr }));
  const prerequisiteMatch = /prerequisite.*:([^:]*)/i.exec(oneLined);
  if (!prerequisiteMatch) throw new Error('could not info out of perquisite block');
  const textContent = prerequisiteMatch[1].trim();
  const withConcurrent = textContent.replace(/__(\w*)\|(\w*)__\s*\*/g, '__$1|$2|CONCURRENT__');
  const withPrevious = withConcurrent.replace(/__(\w*)\|(\w*)__(?!\*)/g, '__$1|$2|PREVIOUS__');
  return withPrevious;
}

/**
 * Recursive function that takes in a string which can include `(` `)` and transforms it into an
 * array of arrays. Strings get separated by spaces and substrings inside of `(` `)` will get put
 * into a sub array:
 *
 * e.g.:
 *
 * ```txt
 * the quick (brown (fox jumps) over the (lazy dog))
 * ```
 *
 * becomes:
 *
 * ```js
 * ['the', 'quick', ['brown', ['fox', 'jumps'], 'over', 'the', ['lazy', 'dog']]]
 * ```
 */
export function transformParenthesesToTree(
  expression: string,
): { tree: ParseTree; lastIndex: number } {
  const characters = expression.split('');
  const tokens = [] as ParseTree;
  let i = 0;
  let currentToken = '';
  while (i < expression.length) {
    const character = characters[i];
    if (character === ' ') {
      // terminal character
      if (currentToken) {
        // check to see if the `currentToken` is falsy
        tokens.push(currentToken);
        currentToken = '';
      }
    } else if (character === '(') {
      const subExpressionResult = transformParenthesesToTree(expression.substring(i + 1));
      i += subExpressionResult.lastIndex;
      tokens.push(subExpressionResult.tree);
    } else if (character === ')') {
      // also terminal character but returns the sub expression
      if (currentToken) {
        // check to see if the `currentToken` is falsy
        tokens.push(currentToken);
      }
      return { tree: tokens, lastIndex: i + 1 };
    } else {
      currentToken += character;
    }
    i += 1;
  }

  if (currentToken) {
    // adds the last current token if there is one
    tokens.push(currentToken);
  }

  return { tree: tokens, lastIndex: i };
}

/**
 * If the string provided matches `/__(.*)\|(.*)__/` (e.g. `blah __CIS|310__ blah`) then the
 * function will return just the `__CIS|310__` and leave out the `blah blah`
 */
export function replaceCourseDirective(token: string) {
  const match = /__(.+)\|(.+)\|(.+)__/.exec(token);
  if (!match) return undefined;

  const subjectCode = match[1].trim().toUpperCase();
  const courseNumber = match[2].trim().toUpperCase();
  const previousOrConcurrent = match[3].trim().toUpperCase();
  if (previousOrConcurrent !== 'PREVIOUS' && previousOrConcurrent !== 'CONCURRENT') {
    throw new Error(
      `Last value in match was neither "PREVIOUS" or "CONCURRENT". It was "${previousOrConcurrent}". Full string "${token}"`,
    );
  }

  return [subjectCode, courseNumber, previousOrConcurrent] as [
    string,
    string,
    'PREVIOUS' | 'CONCURRENT'
  ];
}

export function removeExtraFromDirective(token: string) {
  const result = replaceCourseDirective(token);
  if (!result) return token;

  const [subjectCode, courseNumber, previousOrConcurrent] = result;
  return `__${subjectCode}|${courseNumber}|${previousOrConcurrent}__`;
}

/**
 * Joins strings in an array tree based on whether or not they are operators (i.e. `and` or `or`).
 *
 * transforms:
 * ```
 * ['one', 'two', 'and', ['buckle', 'shoe', 'or', 'three', 'four']]
 * ```
 *
 * into:
 * ```
 * ['one two', 'and', ['buckle shoe', 'or', 'three four']]
 * ```
 */
export function groupByOperator(tree: ParseTree): ParseTree {
  const newTree = [] as ParseTree;
  let currentToken = '';
  for (const node of tree) {
    if (Array.isArray(node)) {
      newTree.push(groupByOperator(node));
    } else if (node.toLowerCase() === 'and' || node.toLowerCase() === 'or') {
      if (currentToken) {
        newTree.push(removeExtraFromDirective(currentToken.trim()));
      }
      newTree.push(node);
      currentToken = '';
    } else {
      currentToken += ' ' + node;
    }
  }
  if (currentToken) {
    newTree.push(removeExtraFromDirective(currentToken.trim()));
  }
  return newTree;
}

function isPrerequisite(obj: any): obj is Prerequisite {
  if (!obj) return false;
  if (typeof obj === 'string') return true;
  if (typeof obj !== 'object') return false;
  if (obj.and) return true;
  if (obj.or) return true;
  if (!Array.isArray(obj)) return false;
  if (obj.length !== 3) return false;
  const last = obj[obj.length - 1];
  if (last === 'CONCURRENT') return true;
  if (last === 'PREVIOUS') return true;
  return false;
}

/**
 * Recursively builds the `_Prerequisite` type given a `ParseTree` from the `tokenizeByOperator`
 * function.
 */
export function buildPrerequisiteTree(tokens: ParseTree | Prerequisite): Prerequisite | undefined {
  if (typeof tokens === 'string') {
    const result = replaceCourseDirective(tokens);
    if (!result) return tokens;
    return result;
  }
  if (isPrerequisite(tokens)) return tokens;

  const transformedTokens = tokens
    .map(token => {
      if (isPrerequisite(token)) return token;
      return buildPrerequisiteTree(token)!;
    })
    .filter(token => !!token);

  const rankedOperators = transformedTokens
    .map((token, index) => ({ token, index }))
    .filter(({ token }) => {
      if (token === 'and') return true;
      if (token === 'or') return true;
      return false;
    })
    .map(({ token, index }) => {
      const baseRank = token === 'and' ? 1000 : 0;
      const rank = baseRank - index;

      return {
        token,
        index,
        rank,
      };
    })
    .sort((a, b) => b.rank - a.rank);

  const bestOperator = rankedOperators[0];
  if (!bestOperator) return undefined;

  const operator = bestOperator.token as 'and' | 'or';
  const previous = transformedTokens[bestOperator.index - 1];
  const next = transformedTokens[bestOperator.index + 1];

  return { [operator]: [previous, next] } as Prerequisite;
}

/**
 * Used to parse both `prerequisites` and `corequisites`. Simply calls all the individual functions.
 */
export function parsePrerequisites(prerequisiteElement: Element) {
  const textContent = replacePrerequisiteAnchors(prerequisiteElement);
  const parseTree = transformParenthesesToTree(textContent);
  const tokens = groupByOperator(parseTree.tree);
  const result = buildPrerequisiteTree(tokens);
  // const result = replaceAllCourseDirectivesInTree(prefix);
  return result;
}
