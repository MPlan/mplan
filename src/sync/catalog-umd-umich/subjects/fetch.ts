import axios from 'axios';

export async function fetchUndergraduateSubjectsHtml() {
  const response = await axios({
    method: 'GET',
    url: 'http://catalog.umd.umich.edu/undergraduate/coursesaz/',
    headers: {
      'X-What-Is-This-Request': 'https://github.com/MPlan/mplan',
    },
  });

  const html = response.data as string | undefined;
  if (!html) {
    throw new Error('Subjects response was undefined or empty.');
  }
  return html;
}
