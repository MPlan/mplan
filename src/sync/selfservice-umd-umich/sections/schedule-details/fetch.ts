import axios from 'axios';
import * as https from 'https';
import { oneLineTrim } from 'common-tags';
import { formEncode } from 'utilities/form-encode-decode';

export async function fetchScheduleDetailHtml(termCode: string, crn: string) {
  const url = oneLineTrim`
    https://selfservice.umd.umich.edu/BANP/bwckschd.p_disp_detail_sched?
    ${formEncode({
      term_in: termCode,
      crn_in: crn,
    })}
  `;

  const response = await axios({
    method: 'POST',
    url,
    httpsAgent: new https.Agent({ ciphers: 'ALL' }),
    headers: {
      'X-What-Is-This-Request': 'https://github.com/MPlan/mplan',
    },
  });

  const html = response.data as string | undefined;
  if (!html) throw new Error('Subjects response was undefined or empty.');

  return html;
}
