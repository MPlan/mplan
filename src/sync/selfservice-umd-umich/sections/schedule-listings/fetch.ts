import * as https from 'https';
import axios from 'axios';
import { oneLineTrim } from 'common-tags';
import { formEncode } from 'utilities/form-encode-decode';

export async function fetchScheduleListingsHtml(
  termCode: string,
  subjectCode: string,
  courseNumber: string,
) {
  const url = oneLineTrim`
    https://selfservice.umd.umich.edu/BANP/bwckctlg.p_disp_listcrse?
    ${formEncode({
      term_in: termCode.toUpperCase().trim(),
      subj_in: subjectCode.toUpperCase().trim(),
      crse_in: courseNumber.toUpperCase().trim(),
      schd_in: '',
    })}
  `;

  const response = await axios({
    method: 'GET',
    url,
    httpsAgent: new https.Agent({ ciphers: 'ALL' }),
    headers: {
      'X-What-Is-This-Request': 'https://github.com/mplan/mplan',
    },
  });

  const html = response.data as string | undefined;
  if (!html) throw new Error('Subjects response was undefined or empty.');

  return html;
}
