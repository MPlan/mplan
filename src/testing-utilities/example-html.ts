import * as fs from 'fs';

// https://stackoverflow.com/questions/16697791/nodejs-get-filename-of-caller-function
function _getCallerFile(): string {
  const originalFunc = Error.prepareStackTrace;

  let callerFile: any;
  try {
    const err = new Error() as any;
    let currentFile;

    Error.prepareStackTrace = function(err, stack) {
      return stack;
    };

    currentFile = err.stack.shift().getFileName();

    while (err.stack.length) {
      callerFile = err.stack.shift().getFileName();
      if (currentFile !== callerFile) break;
    }
  } catch (e) {}
  Error.prepareStackTrace = originalFunc;
  return callerFile;
}

export function writeExampleHtml(id: string, html: string) {
  const callerFile = _getCallerFile();
  const split = callerFile.split('/');
  const callerDirectory = split.slice(0, split.length - 1).join('/');
  const examplesDirectory = `${callerDirectory}/__examples__`;
  if (!fs.existsSync(examplesDirectory)) fs.mkdirSync(examplesDirectory);
  fs.writeFileSync(`${examplesDirectory}/${id}.html`, html);
}

export function getExampleHtml(id: string) {
  const callerFile = _getCallerFile();
  const split = callerFile.split('/');
  const callerDirectory = split.slice(0, split.length - 1).join('/');
  const examplesDirectory = `${callerDirectory}/__examples__`;
  return fs.readFileSync(`${examplesDirectory}/${id}.html`).toString();
}
