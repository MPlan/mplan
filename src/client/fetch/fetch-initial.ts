import { fetchUser } from './user';
import { fetchCatalog } from './catalog';
import { fetchMasteredDegrees } from './mastered-degrees';

export async function fetchInitial() {
  const [user, catalog, masteredDegrees] = await Promise.all([
    fetchUser(),
    fetchCatalog(),
    fetchMasteredDegrees(),
  ]);
  return { user, catalog, masteredDegrees };
}
