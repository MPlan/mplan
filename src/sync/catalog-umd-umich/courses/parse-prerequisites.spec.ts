import { getOrWriteHtml } from 'testing-utilities/example-html';
import { fetchCourseList } from './fetch';
import { parsePrerequisites } from './parse-prerequisites';
import { JSDOM } from 'jsdom';

describe('parsePrerequisites', () => {
  it('works', async () => {
    const html = await getOrWriteHtml('cis-courses', __dirname, () =>
      fetchCourseList('undergraduate', 'cis'),
    );

    const document = new JSDOM(html).window.document;
    const prerequisitesBlocks = Array.from(document.querySelectorAll('.courseblockextra')).filter(
      courseBlockExtra => {
        const textContent = courseBlockExtra.textContent;
        if (!textContent) return false;
        if (!/prerequisite/i.test(textContent)) return false;
        return true;
      },
    );

    const result = prerequisitesBlocks
      .map(block => parsePrerequisites(block))
      .map(result => JSON.stringify(result, null, 2))
      .join('\n====================\n');
    console.log(result);
  });
});
