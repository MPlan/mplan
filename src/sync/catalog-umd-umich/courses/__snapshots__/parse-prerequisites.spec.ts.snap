// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsePrerequisites works 1`] = `
Array [
  Object {
    "or": Array [
      Array [
        "MATH",
        "115",
        "CONCURRENT",
      ],
      Array [
        "MATH",
        "113",
        "CONCURRENT",
      ],
    ],
  },
  Object {
    "or": Array [
      Array [
        "MATH",
        "115",
        "CONCURRENT",
      ],
      Array [
        "MATH",
        "113",
        "CONCURRENT",
      ],
    ],
  },
  "CIS 162",
  Object {
    "and": Array [
      Object {
        "or": Array [
          Array [
            "MATH",
            "115",
            "PREVIOUS",
          ],
          "MPLS with a score of 116",
        ],
      },
      Object {
        "or": Array [
          Array [
            "CIS",
            "150",
            "PREVIOUS",
          ],
          Array [
            "IMSE",
            "150",
            "PREVIOUS",
          ],
        ],
      },
    ],
  },
  Object {
    "and": Array [
      Array [
        "CIS",
        "1501",
        "PREVIOUS",
      ],
      Array [
        "MATH",
        "115",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "or": Array [
      Array [
        "ENGR",
        "100",
        "PREVIOUS",
      ],
      Object {
        "or": Array [
          Array [
            "MATH",
            "105",
            "PREVIOUS",
          ],
          "MPLS with a score of 113",
        ],
      },
    ],
  },
  Object {
    "and": Array [
      Array [
        "MATH",
        "115",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "200",
        "CONCURRENT",
      ],
    ],
  },
  Array [
    "CIS",
    "200",
    "PREVIOUS",
  ],
  Array [
    "CIS",
    "200",
    "CONCURRENT",
  ],
  Array [
    "CIS",
    "200",
    "PREVIOUS",
  ],
  Array [
    "CIS",
    "200",
    "PREVIOUS",
  ],
  Object {
    "or": Array [
      Array [
        "CIS",
        "200",
        "PREVIOUS",
      ],
      Array [
        "IMSE",
        "200",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "or": Array [
      Array [
        "CIS",
        "200",
        "PREVIOUS",
      ],
      Array [
        "IMSE",
        "200",
        "PREVIOUS",
      ],
    ],
  },
  Array [
    "CIS",
    "200",
    "PREVIOUS",
  ],
  Array [
    "CIS",
    "275",
    "PREVIOUS",
  ],
  Object {
    "and": Array [
      Object {
        "or": Array [
          Array [
            "MATH",
            "115",
            "PREVIOUS",
          ],
          "MPLS with a score of 116",
        ],
      },
      Object {
        "or": Array [
          Array [
            "CIS",
            "200",
            "PREVIOUS",
          ],
          Array [
            "IMSE",
            "200",
            "PREVIOUS",
          ],
        ],
      },
    ],
  },
  Object {
    "and": Array [
      Object {
        "or": Array [
          Array [
            "CIS",
            "2001",
            "PREVIOUS",
          ],
          Array [
            "CIS",
            "200",
            "PREVIOUS",
          ],
        ],
      },
      Array [
        "ECE",
        "3100",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "and": Array [
      Object {
        "and": Array [
          Array [
            "MATH",
            "115",
            "PREVIOUS",
          ],
          Object {
            "or": Array [
              Array [
                "CIS",
                "200",
                "PREVIOUS",
              ],
              "MPLS with a score of 116",
            ],
          },
        ],
      },
      Array [
        "CIS",
        "275",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "and": Array [
      Object {
        "or": Array [
          Array [
            "CIS",
            "200",
            "PREVIOUS",
          ],
          Array [
            "IMSE",
            "200",
            "PREVIOUS",
          ],
        ],
      },
      Array [
        "CIS",
        "275",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "and": Array [
      Object {
        "and": Array [
          Array [
            "ECE",
            "370",
            "PREVIOUS",
          ],
          Array [
            "ECE",
            "276",
            "PREVIOUS",
          ],
        ],
      },
      Object {
        "or": Array [
          Array [
            "COMP",
            "270",
            "PREVIOUS",
          ],
          Array [
            "COMP",
            "106",
            "PREVIOUS",
          ],
        ],
      },
    ],
  },
  Array [
    "CIS",
    "375",
    "PREVIOUS",
  ],
  Array [
    "MATH",
    "115",
    "PREVIOUS",
  ],
  Object {
    "and": Array [
      Object {
        "or": Array [
          Array [
            "CIS",
            "200",
            "PREVIOUS",
          ],
          Array [
            "ECE",
            "270",
            "PREVIOUS",
          ],
        ],
      },
      Object {
        "or": Array [
          Array [
            "CIS",
            "310",
            "CONCURRENT",
          ],
          Array [
            "ECE",
            "370",
            "CONCURRENT",
          ],
        ],
      },
    ],
  },
  Object {
    "or": Array [
      Array [
        "CIS",
        "350",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "3501",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "or": Array [
      Array [
        "CIS",
        "350",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "3501",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "or": Array [
      Object {
        "or": Array [
          Array [
            "CIS",
            "350",
            "PREVIOUS",
          ],
          Array [
            "IMSE",
            "350",
            "PREVIOUS",
          ],
        ],
      },
      Object {
        "and": Array [
          Array [
            "ECE",
            "370",
            "PREVIOUS",
          ],
          Array [
            "MATH",
            "276",
            "PREVIOUS",
          ],
        ],
      },
    ],
  },
  Array [
    "CIS",
    "350",
    "PREVIOUS",
  ],
  Object {
    "or": Array [
      Array [
        "CIS",
        "350",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "3501",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "and": Array [
      Object {
        "or": Array [
          Array [
            "CIS",
            "2001",
            "PREVIOUS",
          ],
          Array [
            "CIS",
            "200",
            "PREVIOUS",
          ],
        ],
      },
      Array [
        "CIS",
        "3200",
        "PREVIOUS",
      ],
    ],
  },
  Array [
    "CIS",
    "421",
    "PREVIOUS",
  ],
  Object {
    "and": Array [
      Array [
        "CIS",
        "375",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "421",
        "CONCURRENT",
      ],
    ],
  },
  Object {
    "and": Array [
      Array [
        "CIS",
        "375",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "421",
        "CONCURRENT",
      ],
    ],
  },
  Array [
    "CIS",
    "4261",
    "PREVIOUS",
  ],
  Object {
    "and": Array [
      Object {
        "and": Array [
          Object {
            "and": Array [
              Array [
                "ECE",
                "370",
                "PREVIOUS",
              ],
              Array [
                "ECE",
                "276",
                "PREVIOUS",
              ],
            ],
          },
          Array [
            "MATH",
            "276",
            "PREVIOUS",
          ],
        ],
      },
      Array [
        "IMSE",
        "317",
        "PREVIOUS",
      ],
    ],
  },
  Array [
    "CIS",
    "427",
    "PREVIOUS",
  ],
  Array [
    "CIS|553|PREVIOUS__ __CIS",
    "375",
    "CONCURRENT",
  ],
  Array [
    "CIS",
    "375",
    "CONCURRENT",
  ],
  Array [
    "CIS",
    "427",
    "PREVIOUS",
  ],
  Array [
    "CIS",
    "450",
    "CONCURRENT",
  ],
  Object {
    "and": Array [
      Array [
        "CIS",
        "310",
        "PREVIOUS",
      ],
      Object {
        "or": Array [
          Array [
            "CIS",
            "350",
            "PREVIOUS",
          ],
          Array [
            "CIS",
            "3501",
            "PREVIOUS",
          ],
        ],
      },
    ],
  },
  Array [
    "MATH",
    "227",
    "PREVIOUS",
  ],
  Object {
    "or": Array [
      Array [
        "CIS",
        "451",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "487",
        "PREVIOUS",
      ],
    ],
  },
  Array [
    "CIS",
    "350",
    "PREVIOUS",
  ],
  Object {
    "and": Array [
      Object {
        "or": Array [
          Array [
            "CIS",
            "427",
            "CONCURRENT",
          ],
          Array [
            "ECE",
            "471",
            "CONCURRENT",
          ],
        ],
      },
      Object {
        "or": Array [
          Array [
            "CIS",
            "387",
            "PREVIOUS",
          ],
          Array [
            "ECE",
            "387",
            "PREVIOUS",
          ],
        ],
      },
    ],
  },
  Object {
    "or": Array [
      Array [
        "CIS",
        "350",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "3501",
        "PREVIOUS",
      ],
    ],
  },
  Array [
    "CIS",
    "376",
    "PREVIOUS",
  ],
  Array [
    "CIS",
    "375",
    "PREVIOUS",
  ],
  Object {
    "or": Array [
      Array [
        "CIS",
        "350",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "3501",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "and": Array [
      Array [
        "CIS",
        "306",
        "PREVIOUS",
      ],
      Object {
        "or": Array [
          Array [
            "MATH",
            "217",
            "CONCURRENT",
          ],
          Array [
            "MATH",
            "227",
            "CONCURRENT",
          ],
        ],
      },
    ],
  },
  Object {
    "or": Array [
      Array [
        "CIS",
        "200",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "2001",
        "PREVIOUS",
      ],
    ],
  },
  Array [
    "CIS",
    "375",
    "CONCURRENT",
  ],
  Array [
    "CIS",
    "487",
    "PREVIOUS",
  ],
  Object {
    "or": Array [
      Array [
        "CIS",
        "350",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "3501",
        "PREVIOUS",
      ],
    ],
  },
  Object {
    "and": Array [
      Object {
        "or": Array [
          Array [
            "CIS",
            "200",
            "PREVIOUS",
          ],
          Array [
            "CIS",
            "2001",
            "PREVIOUS",
          ],
        ],
      },
      Array [
        "MATH",
        "396",
        "PREVIOUS",
      ],
    ],
  },
  Array [
    "CIS",
    "375",
    "PREVIOUS",
  ],
  Array [
    "CIS",
    "375",
    "PREVIOUS",
  ],
  Array [
    "CIS",
    "4951",
    "PREVIOUS",
  ],
  Object {
    "and": Array [
      Array [
        "CIS",
        "376",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "476",
        "PREVIOUS",
      ],
    ],
  },
  Array [
    "CIS",
    "376",
    "PREVIOUS",
  ],
  Object {
    "and": Array [
      Array [
        "CIS",
        "4961",
        "PREVIOUS",
      ],
      Array [
        "CIS",
        "476",
        "CONCURRENT",
      ],
    ],
  },
  Object {
    "and": Array [
      Array [
        "CIS",
        "3200",
        "PREVIOUS",
      ],
      Object {
        "or": Array [
          Array [
            "STAT",
            "326",
            "PREVIOUS",
          ],
          "STAT 425",
        ],
      },
    ],
  },
  Object {
    "and": Array [
      Array [
        "CIS",
        "4971",
        "PREVIOUS",
      ],
      Array [
        "STAT",
        "430",
        "CONCURRENT",
      ],
    ],
  },
]
`;
